### DISCUSSION TITLE
Question: extending pocketbase

### DISCUSSION THREAD

**Discussion by GuillaumeQuenneville** [pocketbase/pocketbase#1222] [Q&A] [answered]
I'd like to add multiple event handlers and routes to to my pocketbase.  Enough to want to put them in separate files and/or packages.

All the examples for updating records requires the app instance but the Record events don't have a reference to the app (e.g. https://pocketbase.io/docs/record-methods/#update-record-without-data-validations) . At the moment I wrap all my event handlers with the app but is this the proper way of doing it? 

I am new to Go. Is there an example of a larger project with custom routes and handlers that are well organized to study? I've had poor luck getting help from other channels.

I'd be happy to submit new/update the documentation when I have an answer if you think it would be useful.

```go
package main

import (
	"log"

	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/core"
	"github.com/pocketbase/pocketbase/models"
)

func doStuff(app *pocketbase.PocketBase) func(e *core.RecordCreateEvent) error {
	return func(e *core.RecordCreateEvent) error {
		if e.Record.Collection().Name == "jobs" {
			// start goroutine to do long running task
			go func(record *models.Record) {
				record.Set("status", "running")
				if err := app.Dao().SaveRecord(record); err != nil {
					// handle error
				}
				// Do work
				// ...
                                time.Sleep(10 * time.Second)
				record.Set("status", "completed")
				if err := app.Dao().SaveRecord(record); err != nil {
					// handle error
				}
			}(e.Record)
		}
		return nil
	}
}

func main() {
	app := pocketbase.New()
	app.OnRecordAfterCreateRequest().Add(
		doStuff(app),
	)
	if err := app.Start(); err != nil {
		log.Fatal(err)
	}
}
```

**Comment by ganigeorgiev** [comment, marked as answer]
> At the moment I wrap all my event handlers with the app but is this the proper way of doing it?

Yes, it is fine. In the future I'll consider passing the `core.App` instance as part of the event hook data to make it a little more easier to compose.
